pkg/pidfile/pidfile.go

```diff
@@ -7,36 +7,50 @@ import (
 	"bytes"
 	"fmt"
 	"os"
-	"path/filepath"
 	"strconv"
 
-	"github.com/docker/docker/pkg/system"
+	"github.com/docker/docker/pkg/process"
 )
 
-func checkPIDFileAlreadyExists(path string) error {
+func readPIDFile(path string) (pid int, err error) {
 	pidByte, err := os.ReadFile(path)
 	if err != nil {
-		if os.IsNotExist(err) {
-			return nil
-		}
-		return err
+		return 0, err
+	}
+	pid, err = strconv.Atoi(string(bytes.TrimSpace(pidByte)))
+	if err != nil {
+		return 0, nil
 	}
-	pid, err := strconv.Atoi(string(bytes.TrimSpace(pidByte)))
-	if err == nil && processExists(pid) {
-		return fmt.Errorf("pid file found, ensure docker is not running or delete %s", path)
+	if pid != 0 && process.Alive(pid) {
+		return pid, nil
 	}
-	return nil
+	return 0, nil
 }
 
 // Write writes a "PID file" at the specified path. It returns an error if the
 // file exists and contains a valid PID of a running process, or when failing
 // to write the file.
-func Write(path string) error {
-	if err := checkPIDFileAlreadyExists(path); err != nil {
-		return err
+func Write(path string, pid int) error {
+	if pid < 1 {
+		// We might be running as PID 1 when running docker-in-docker,
+		// but 0 or negative PIDs are not acceptable.
+		return fmt.Errorf("invalid PID: %d", pid)
 	}
-	if err := system.MkdirAll(filepath.Dir(path), 0o755); err != nil {
+	oldPID, err := readPIDFile(path)
+	if err != nil && !os.IsNotExist(err) {
 		return err
 	}
-	return os.WriteFile(path, []byte(strconv.Itoa(os.Getpid())), 0o644)
+	if oldPID != 0 {
+		return fmt.Errorf("process with PID %d is still running", oldPID)
+	}
+	return os.WriteFile(path, []byte(strconv.Itoa(pid)), 0o644)
+}
+
+// Read reads the "PID file" at path, and returns the PID if it contains a
+// valid PID of a running process, or 0 otherwise. It returns an error when
+// failing to read the file, or if the file doesn't exist, but malformed content
+// is ignored. Consumers should therefore check if the returned PID is a non-zero
+// value before use.
+func Read(path string) (pid int, err error) {
+	return readPIDFile(path)
 }
```

Comment:
Why write a trivial wrapper function instead of simply renaming `readPIDFile` to `Read`? The signatures are identical so the wrapper is not doing any implicit type-conversions.

Reply:
Looks like there's no reason indeed (thought maybe we were differentiating per platform, but we weren't, so updated it 👍